mods/bucket/init.lua

```diff
@@ -153,6 +153,22 @@ minetest.register_craftitem("bucket:bucket_empty", {
 			minetest.add_node(pointed_thing.under, {name="air"})
 
 			return ItemStack(giving_back)
+		else
+			local node_def = minetest.registered_nodes[node.name]
+			if node_def then
+				-- Buckets will run a node's on_punch function if it is not liquid.
+				if node_def.on_punch then
+					giving_back = node_def.on_punch(
+							pointed_thing.under,
+							minetest.get_node(pointed_thing.under),
+							user,
+							pointed_thing)
+					
+					if giving_back ~= nil then
+						return ItemStack(giving_back)
+					end
```

Comment:
You should remove this block. You cannot rely on the return value being ItemStack compatible and it might crash, when someone returns stuff one day (like from an api addition).

Instead you shall always return the leftover itemstack. And that one is already passed to the function:
```lua
return itemstack
```